src/tokenization/bpe_tokenizer.py

```
@@ -183,107 +245,101 @@ def train(self, texts: List[str], verbose: bool = True):
+            if best_pair[3] == float("inf"):
                 # No more valid merges
                 break
 
             # Merge this pair
-            pair_to_merge = (best_pair[0], best_pair[1])
-            new_token = self.merges[pair_to_merge]
+            i, first, second, _ = best_pair
+            new_token = self.merges[(first, second)]
+
+            # Rebuild tokens with merge applied
+            tokens = tokens[:i] + [new_token] + tokens[i+2:]
```

Ruff (0.14.2)

Consider `[*tokens[:i], new_token, *tokens[i + 2:]]` instead of concatenation

Replace with `[*tokens[:i], new_token, *tokens[i + 2:]]`

(RUF005)